Fixes for platforms where init container fails to run

aditighag opened the pull request:
See commit messages.

Fixes: #16814 / #16828
Fixes: #16771
Fixes: fa8bea4 ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

**Release note**
```release-note
Removes cilium daemonset's dependencies on utilities like `sh` and `mount` having installed in the underlying host distributions.
```

aditighag committed:
install/kubernetes: Mount host's `/proc` in init container

The mount-cgroup init container needs to run a mount command
on the underlying host. But the current approach to mount
`/proc/1/ns` fails on distros like Fedora when running on
 minikube -

mounting "/proc/1/ns" to rootfs at "/var/lib/docker/.../merged/hostpid1ns" caused: permission denied: unknown

Mount host's `/proc` instead.

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Reported-By: André Martins <andre@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
install/kubernetes: Set image pull policy for init container

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -384,16 +384,16 @@ spec:
             value: {{ .Values.cgroup.hostRoot }}
```
please use `Reported-By: André Martins <andre@cilium.io>` and add `Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")` in the commit message

aditighag replied:
>please use Reported-By:

Ah, I copied this from your cilium commits. Will update the email id.

aanm commented on mount/main.go:
```diff
@@ -0,0 +1,70 @@
+//  Copyright 2021 Authors of Cilium
```
maybe this (mount utility) could live under `tools/` directory?

aditighag replied:
I've tried to follow the same model as bugtool as discussed on the Slack channel. The reason being we don't seem to create binaries in `/usr/bin/cilium` for any of the `/tools/*` programs. AFAICT, those programs are invoked directly using `go run`.

aanm replied:
Yes but this will start cluttering the root repository. Although those programs are invoked using `go run` it should not be a requirement for them to live under `/tools`. `bugtool` should likely be under that directory as well.

aditighag replied:
Yeah, that's a fair point. I'll move it under `tools`.

aanm commented on mount/main.go:
```diff
@@ -0,0 +1,70 @@
+		os.Exit(1)
+	}
+
+	cgroupRoot := os.Args[1]
```
why can't we simply call `cgroups.CheckOrMountCgrpFS(cgroupRoot)`?

aditighag replied:
We could; it's more or less the same code. But the logic is faulty so we'll need to first merge - https://github.com/cilium/cilium/pull/15999. I also prefer to avoid duplicating the code. Also, the function currently doesn't return an error so we'll need to fix that.

I made one more change to this function - when the passed path has a mount, but different fs type, we currently return error. But that might not be always true. For example, if cgroup v1 fs is mounted, we can still mount cgroup2 fs there. Given that the auto mount path is set to /run/cilium/cgroupv2, this change isn't necessarily required though. 

Edit : I'm not sure we want to use the exact same function since it has other dependencies like logging, etc.

aanm replied:
PR #15999 should be merged by then.

>Edit : I'm not sure we want to use the exact same function since it has other dependencies like logging, etc.

it's fine, it will make the logging consistent across cilium binaries.

aditighag replied:
See this comment - https://github.com/cilium/cilium/pull/16815#discussion_r665047791. I'm leaning towards keeping the code executed by this program to as minimal as possible. Let's see what @joestringer has to say as well. Furthermore, if there are issues with the mount point directory, the mount syscall should print an error. 

>$ sudo mount -t cgroup2 none /var/run/test
mount: /var/run/test: mount point does not exist.

joestringer replied:
Six of one, half dozen of the other. It looks to me like lines 59-68 are trying to achieve the same logic as `cgrpCheckOrMountLocation()` so we may as well have exactly one copy of that logic and make sure it's the way we want it for both use cases.

aditighag commented on mount/main.go:
```diff
@@ -0,0 +1,70 @@
+		}
+	} else if !cgroupRootStat.IsDir() {
+		return fmt.Errorf("%s is a file which is not a directory", cgroupRoot)
+	}
```
I think we can get rid of this code. We'll be creating the directory if not present from cilium agent daemonset.

aditighag commented on mount/Makefile:
```diff
@@ -0,0 +1,27 @@
+# Copyright 2017-2020 Authors of Cilium
```
I've fixed the copyright.

joestringer commented on mount/main.go:
```diff
@@ -0,0 +1,70 @@
+	cgroupRoot := os.Args[1]
+	mounted, cgroupInstance, err := mountinfo.IsMountFS(mountinfo.FilesystemTypeCgroup2, cgroupRoot)
+	if err != nil {
+		fmt.Fprintf(os.Stdout, "Failed to check mount point %s\n", cgroupRoot)
```
Errors should be printed to `os.Stderr`.

aditighag committed:
install/kubernetes: Remove `sh` and `mount` dependency from init container

The mount-cgroup init container runs a mount command on the underlying
host using `nsenter`. However, certain distros like Talos don't have `sh`
or `mount` utilities available. Hence, move the logic to check and mount
cgroup2 fs to a statically linked Go program binary.

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -382,18 +382,24 @@ spec:
+          # distros installed on the underlying host. Copy the binary to the
+          # same directory where we install cilium cni plugin so that exec permissions
+          # are available.
+          - 'cp /usr/bin/cilium-mount /hostbin/cilium-mount && nsenter --cgroup=/hostproc/1/ns/cgroup --mount=/hostproc/1/ns/mnt $BIN_PATH/cilium-mount $CGROUP_ROOT; rm /hostbin/cilium-mount'
```
`"${BIN_PATH}/cilium-mount"` to prevent word splitting.

aanm reviewed (approved):
One small nit, doesn't require my 2nd review.

joestringer reviewed (approved):
LGTM apart from @aanm's nit. It's generally better to follow @aanm's suggestion around quoting variables in shell scripts to prevent unexpected argument handling, even though I don't expect that it will have unexpected impacts in this specific case.
